File: app/frontend/core_external/models/lightweight_layers/layers_pooling.js
/**
 * Created by ar on 21.01.17.
 */

import {LW_Layer, default_dim_ordering} from './layers_basic';
import {conv_output_length} from './layers_convolutional';


//////////////////////////////////////////////
class _LW_Pooling1D extends LW_Layer {
    constructor({pool_length=2, stride=null, border_mode='valid'}) {
        super();
        this.input_dim = 3;
        if (stride===null) {
            stride = pool_length;
        }
        if(['valid', 'same'].indexOf(border_mode)<0) {
            throw new TypeError(`'border_mode must be in {valid, same}'`);
        }
        this.pool_length    = pool_length;
        this.stride         = stride;
        this.border_mode    = border_mode;
    }
    get_output_shape_for(input_shape) {
        let lengthOut = conv_output_length({
            input_length:   input_shape[1],
            filter_size:    this.pool_length,
            border_mode:    this.border_mode,
            stride:         this.stride
        });
        return [input_shape[0], lengthOut, input_shape[2]]
    }
}

export class LW_MaxPooling1D extends _LW_Pooling1D {
    constructor({pool_length=2, stride=null, border_mode='valid'}) {
        super({
            pool_length:    pool_length,
            stride:         stride,
            border_mode:    border_mode
        });
    }
}

export class LW_AveragePooling1D extends _LW_Pooling1D {
    constructor({pool_length=2, stride=null, border_mode='valid'}) {
        super({
            pool_length:    pool_length,
            stride:         stride,
            border_mode:    border_mode
        });
    }
}

//////////////////////////////////////////////
class _LW_Pooling2D extends LW_Layer {
    constructor({pool_size=[2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super();
        if (dim_ordering === 'default') {
            dim_ordering = default_dim_ordering;
        }
        if(['tf', 'th'].indexOf(dim_ordering)<0) {
            throw new TypeError('dim_ordering must be in {tf, th}');
        }
        this.pool_size = pool_size;
        if (strides===null) {
            strides = this.pool_size;
        }
        if(['valid', 'same'].indexOf(border_mode)<0) {
            throw new TypeError(`'border_mode must be in {valid, same}'`);
        }
        this.strides        = strides;
        this.border_mode    = border_mode;
        this.dim_ordering   = dim_ordering;
    }
    get_output_shape_for(input_shape) {
        let rows, cols;
        if (this.dim_ordering === 'th') {
            rows = input_shape[2];
            cols = input_shape[3];
        } else if (this.dim_ordering === 'tf') {
            rows = input_shape[1];
            cols = input_shape[2];
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
        rows = conv_output_length({
            input_length:   rows,
            filter_size:    this.pool_size[0],
            border_mode:    this.border_mode,
            stride:         this.strides[0]
        });
        cols = conv_output_length({
            input_length:   cols,
            filter_size:    this.pool_size[1],
            border_mode:    this.border_mode,
            stride:         this.strides[1]
        });
        if (this.dim_ordering === 'th') {
            return [input_shape[0], input_shape[1], rows, cols];
        } else if (this.dim_ordering === 'tf') {
            return [input_shape[0], rows, cols, input_shape[3]];
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
    }
}

export class LW_MaxPooling2D extends _LW_Pooling2D {
    constructor({pool_size=[2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super({
            pool_size:      pool_size,
            strides:        strides,
            border_mode:    border_mode,
            dim_ordering:   dim_ordering
        });
    }
}

export class LW_AveragePooling2D extends _LW_Pooling2D {
    constructor({pool_size=[2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super({
            pool_size:      pool_size,
            strides:        strides,
            border_mode:    border_mode,
            dim_ordering:   dim_ordering
        });
    }
}

//////////////////////////////////////////////
class _LW_Pooling3D extends LW_Layer {
    constructor({pool_size=[2, 2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super();
        if (dim_ordering === 'default') {
            dim_ordering = default_dim_ordering;
        }
        if(['tf', 'th'].indexOf(dim_ordering)<0) {
            throw new TypeError('dim_ordering must be in {tf, th}');
        }
        this.pool_size = pool_size;
        if (strides===null) {
            strides = this.pool_size;
        }
        this.strides = strides;
        if(['valid', 'same'].indexOf(border_mode)<0) {
            throw new TypeError(`'border_mode must be in {valid, same}'`);
        }
        this.border_mode    = border_mode;
        this.dim_ordering   = dim_ordering;
    }
    get_output_shape_for(input_shape) {
        let len_dim1, len_dim2, len_dim3;
        if (this.dim_ordering === 'th') {
            len_dim1 = input_shape[2];
            len_dim2 = input_shape[3];
            len_dim3 = input_shape[4];
        } else if (this.dim_ordering === 'tf') {
            len_dim1 = input_shape[1];
            len_dim2 = input_shape[2];
            len_dim3 = input_shape[3];
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
        len_dim1 = conv_output_length({
            input_length:   len_dim1,
            filter_size:    this.pool_size[0],
            border_mode:    this.border_mode,
            stride:         this.strides[0]
        });
        len_dim2 = conv_output_length({
            input_length:   len_dim2,
            filter_size:    this.pool_size[1],
            border_mode:    this.border_mode,
            stride:         this.strides[1]
        });
        len_dim3 = conv_output_length({
            input_length:   len_dim3,
            filter_size:    this.pool_size[2],
            border_mode:    this.border_mode,
            stride:         this.strides[2]
        });
        if (this.dim_ordering === 'th') {
            return [input_shape[0], input_shape[1], len_dim1, len_dim2, len_dim3];
        } else if (this.dim_ordering === 'tf') {
            return [input_shape[0], len_dim1, len_dim2, len_dim3, input_shape[4]];
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
    }
}

export class LW_MaxPooling3D extends _LW_Pooling3D {
    constructor({pool_size=[2, 2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super({
            pool_size:      pool_size,
            strides:        strides,
            border_mode:    border_mode,
            dim_ordering:   dim_ordering
        });
    }
}

export class LW_AveragePooling3D extends _LW_Pooling3D {
    constructor({pool_size=[2, 2, 2], strides=null, border_mode='valid', dim_ordering='default'}) {
        super({
            pool_size:      pool_size,
            strides:        strides,
            border_mode:    border_mode,
            dim_ordering:   dim_ordering
        });
    }
}

//////////////////////////////////////////////
class _LW_GlobalPooling1D extends LW_Layer {
    constructor() {
        super();
    }
    get_output_shape_for(input_shape) {
        return [input_shape[0], input_shape[2]];
    }
}

export class LW_GlobalAveragePooling1D extends _LW_GlobalPooling1D {
    constructor() {
        super();
    }
}

export class LW_GlobalMaxPooling1D extends _LW_GlobalPooling1D {
    constructor() {
        super();
    }
}

//////////////////////////////////////////////
class _LW_GlobalPooling2D extends LW_Layer {
    constructor(dim_ordering='default') {
        super();
        if (dim_ordering === 'default') {
            dim_ordering = default_dim_ordering;
        }
        this.dim_ordering = dim_ordering;
    }
    get_output_shape_for(input_shape) {
        if(this.dim_ordering==='tf') {
            return [input_shape[0], input_shape[3]];
        } else if(this.dim_ordering==='th') {
            return [input_shape[0], input_shape[1]]
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
    }
}

export class LW_GlobalAveragePooling2D extends _LW_GlobalPooling2D {
    constructor() {
        super();
    }
}

export class LW_GlobalMaxPooling2D extends _LW_GlobalPooling2D {
    constructor() {
        super();
    }
}

//////////////////////////////////////////////
class _LW_GlobalPooling3D extends LW_Layer {
    constructor(dim_ordering='default') {
        super();
        if (dim_ordering === 'default') {
            dim_ordering = default_dim_ordering;
        }
        this.dim_ordering = dim_ordering;
    }
    get_output_shape_for(input_shape) {
        if(this.dim_ordering==='tf') {
            return [input_shape[0], input_shape[4]];
        } else if(this.dim_ordering==='th') {
            return [input_shape[0], input_shape[1]]
        } else {
            throw new TypeError(`Invalid dim_ordering: ${this.dim_ordering}`);
        }
    }
}

export class LW_GlobalAveragePooling3D extends _LW_GlobalPooling3D {
    constructor() {
        super();
    }
}

export class LW_GlobalMaxPooling3D extends _LW_GlobalPooling3D {
    constructor() {
        super();
    }
}

File: app/backend/core/models-keras-2x-api/lightweight_layers/layers_pooling.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
__author__ = 'ar'

from layers_basic import LW_Layer, default_data_format
from layers_convolutional import conv_output_length

###############################################
class _LW_Pooling1D(LW_Layer):
    input_dim = 3
    def __init__(self, pool_size=2, strides=None, padding='valid'):
        if strides is None:
            strides = pool_size
        assert padding in {'valid', 'same'}, 'border_mode must be in {valid, same}'
        self.pool_length = pool_size
        self.stride = strides
        self.border_mode = padding
    def get_output_shape_for(self, input_shape):
        length = conv_output_length(input_shape[1], self.pool_length, self.border_mode, self.stride)
        return (input_shape[0], length, input_shape[2])

class LW_MaxPooling1D(_LW_Pooling1D):
    def __init__(self, pool_size=2, strides=None, padding='valid'):
        super(LW_MaxPooling1D, self).__init__(pool_size, strides, padding)

class LW_AveragePooling1D(_LW_Pooling1D):
    def __init__(self, pool_size=2, strides=None, padding='valid'):
        super(LW_AveragePooling1D, self).__init__(pool_size, strides, padding)

###############################################
class _LW_Pooling2D(LW_Layer):
    def __init__(self, pool_size=(2, 2), strides=None, padding='valid', data_format='default'):
        if data_format == 'default':
            data_format = default_data_format
        assert data_format in {'channels_last', 'channels_first'}, 'data_format must be in {channels_last, channels_first}'
        self.pool_size = tuple(pool_size)
        if strides is None:
            strides = self.pool_size
        self.strides = tuple(strides)
        assert padding in {'valid', 'same'}, 'border_mode must be in {valid, same}'
        self.border_mode = padding
        self.dim_ordering = data_format
    def get_output_shape_for(self, input_shape):
        if self.dim_ordering == 'channels_first':
            rows = input_shape[2]
            cols = input_shape[3]
        elif self.dim_ordering == 'channels_last':
            rows = input_shape[1]
            cols = input_shape[2]
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
        rows = conv_output_length(rows, self.pool_size[0], self.border_mode, self.strides[0])
        cols = conv_output_length(cols, self.pool_size[1], self.border_mode, self.strides[1])
        if self.dim_ordering == 'channels_first':
            return (input_shape[0], input_shape[1], rows, cols)
        elif self.dim_ordering == 'channels_last':
            return (input_shape[0], rows, cols, input_shape[3])
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)

class LW_MaxPooling2D(_LW_Pooling2D):
    def __init__(self, pool_size=(2, 2), strides=None, padding='valid', data_format='default'):
        super(LW_MaxPooling2D, self).__init__(pool_size, strides, padding, data_format)

class LW_AveragePooling2D(_LW_Pooling2D):
    def __init__(self, pool_size=(2, 2), strides=None, padding='valid', data_format='default'):
        super(LW_AveragePooling2D, self).__init__(pool_size, strides, padding, data_format)

###############################################
class _LW_Pooling3D(LW_Layer):
    def __init__(self, pool_size=(2, 2, 2), strides=None, border_mode='valid', dim_ordering='default'):
        if dim_ordering == 'default':
            dim_ordering = default_data_format
        assert dim_ordering in {'channels_last', 'channels_first'}, 'data_format must be in {channels_last, channels_first}'
        self.pool_size = tuple(pool_size)
        if strides is None:
            strides = self.pool_size
        self.strides = tuple(strides)
        assert border_mode in {'valid', 'same'}, 'border_mode must be in {valid, same}'
        self.border_mode = border_mode
        self.dim_ordering = dim_ordering
    def get_output_shape_for(self, input_shape):
        if self.dim_ordering == 'channels_first':
            len_dim1 = input_shape[2]
            len_dim2 = input_shape[3]
            len_dim3 = input_shape[4]
        elif self.dim_ordering == 'channels_last':
            len_dim1 = input_shape[1]
            len_dim2 = input_shape[2]
            len_dim3 = input_shape[3]
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
        len_dim1 = conv_output_length(len_dim1, self.pool_size[0], self.border_mode, self.strides[0])
        len_dim2 = conv_output_length(len_dim2, self.pool_size[1], self.border_mode, self.strides[1])
        len_dim3 = conv_output_length(len_dim3, self.pool_size[2], self.border_mode, self.strides[2])
        if self.dim_ordering == 'channels_first':
            return (input_shape[0], input_shape[1], len_dim1, len_dim2, len_dim3)
        elif self.dim_ordering == 'channels_last':
            return (input_shape[0], len_dim1, len_dim2, len_dim3, input_shape[4])
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)

class LW_MaxPooling3D(_LW_Pooling3D):
    def __init__(self, pool_size=(2, 2, 2), strides=None, border_mode='valid', dim_ordering='default'):
        super(LW_MaxPooling3D, self).__init__(pool_size, strides, border_mode, dim_ordering)

class LW_AveragePooling3D(_LW_Pooling3D):
    def __init__(self, pool_size=(2, 2, 2), strides=None, border_mode='valid', dim_ordering='default'):
        super(LW_AveragePooling3D, self).__init__(pool_size, strides, border_mode, dim_ordering)

###############################################
class _LW_GlobalPooling1D(LW_Layer):
    def __init__(self):
        pass
    def get_output_shape_for(self, input_shape):
        return (input_shape[0], input_shape[2])

class LW_GlobalAveragePooling1D(_LW_GlobalPooling1D):
    pass

class LW_GlobalMaxPooling1D(_LW_GlobalPooling1D):
    pass

###############################################
class _LW_GlobalPooling2D(LW_Layer):

    def __init__(self, data_format='default'):
        if data_format == 'default':
            data_format = default_data_format
        self.dim_ordering = data_format
    def get_output_shape_for(self, input_shape):
        if self.dim_ordering == 'channels_last':
            return (input_shape[0], input_shape[3])
        else:
            return (input_shape[0], input_shape[1])

class LW_GlobalAveragePooling2D(_LW_GlobalPooling2D):
    pass

class LW_GlobalMaxPooling2D(_LW_GlobalPooling2D):
    pass

###############################################
class _LW_GlobalPooling3D(LW_Layer):
    def __init__(self, data_format='default'):
        if data_format == 'default':
            data_format = default_data_format
        self.dim_ordering = data_format
    def get_output_shape_for(self, input_shape):
        if self.dim_ordering == 'channels_last':
            return (input_shape[0], input_shape[4])
        else:
            return (input_shape[0], input_shape[1])

class LW_GlobalAveragePooling3D(_LW_GlobalPooling3D):
    pass

class LW_GlobalMaxPooling3D(_LW_GlobalPooling3D):
    pass

###############################################
if __name__ == '__main__':
    pass